[p1] CI flake: SwiftShims module cache `mtime` mismatch fails Build & Test on every PR run

## Summary

Multiple PRs (#716, #724, #717, #723, #733 — every PR opened against `main` over the past ~6 hours) are failing the `Build & Test` job with the same Swift compiler error before any test can run:

```
<unknown>:0: error: file '/Applications/Xcode_16.4.app/Contents/Developer/Platforms/iPhoneSimulator.platform/Developer/SDKs/iPhoneSimulator18.5.sdk/usr/lib/swift/shims/module.modulemap' has been modified since the module file '/Users/runner/work/fantastic-octo-fortnight/fantastic-octo-fortnight/DerivedData/ModuleCache.noindex/<runner-id>/SwiftShims-<hash>.pcm' was built: mtime changed (was 1778476540, now 1777267324)
** BUILD FAILED **
Error: ... Failed to load result bundle
::error::TestResults.xcresult not found — tests may have crashed before producing results
::error::Coverage JSON was empty from xccov
```

(See e.g. https://github.com/yashasg/fantastic-octo-fortnight/actions/runs/25905913514/job/76141655824 for PR #716, https://github.com/yashasg/fantastic-octo-fortnight/actions/runs/25905852039/job/76140000505 for PR #724.)

## Root cause hypothesis

GitHub Actions `macos-latest` runners share the macOS SDK installation across concurrent jobs, but each job has its own `DerivedData/ModuleCache.noindex` populated by an `actions/cache` (or similar) restore. When the SDK `module.modulemap` `mtime` differs from what the cached `.pcm` was built against, Swift refuses to use the cache and the build fails before the test phase.

The `(was 1778476540, now 1777267324)` ordering — "was" newer than "now" — suggests the cached module file came from a runner that had a *more recent* SDK `mtime` than the runner currently restoring it. This rules out a one-way upgrade and implicates the cross-runner mtime variance that GHA `macos-15` images currently exhibit.

## Reproduction

100% reproduction: any PR opened against `main` since 2026-05-15 ~07:00Z. Previously-green commits (e.g. main HEAD `72ca78a`) succeeded, so this is environmental, not a code regression.

## Workaround (current)

Force-pushing a rebase onto current main triggers a fresh CI run that *sometimes* lands on a runner whose SDK mtime matches the cache. Inconsistent — succeeds ~30% of attempts.

## Required fix (proposal)

One of:

**A. Disable Swift module cache on CI** — pass `-Xfrontend -no-clang-module-breadcrumbs` and/or set `SWIFT_DISABLE_MODULE_CACHE_VALIDATION=YES` (or equivalent `xcodebuild` setting) in `scripts/build.sh` when `CI=true`. Eliminates the mtime check at the cost of slower builds.

**B. Skip the cache-key restore for the SDK-shim module** — narrow the `actions/cache` key in `.github/workflows/build.yml` to exclude `DerivedData/ModuleCache.noindex/**/SwiftShims-*.pcm`. Keeps cache benefit for app modules.

**C. Pin to a fixed Xcode version** — replace `macos-latest` with `macos-15` + explicit `xcode-select` to a stable Xcode build, removing the cross-runner variance entirely.

Option A is the smallest change and directly addresses the root cause; Option C is the most durable.

## Acceptance criteria

1. `Build & Test` succeeds on a freshly opened PR against `main` 5/5 consecutive runs without any rebase or re-trigger.
2. Existing 2208-test suite remains green.
3. Build wall-clock time on CI does not regress more than 30% (Option A risk).

## Priority rationale

p1: blocks the entire #677 / #701 / #702 p0-tca-14* chain from merging, plus #680, #678, #711, #732. Currently every PR has to be re-triggered 1-3× before it lands a non-flaky run, which is burning ~30 minutes of CI per PR and stalling all in-flight squad work.

## Refs

- Blocks #677, #701, #702, #680, #678, #711, #732
- PRs currently affected: #716, #717, #723, #724, #733

--- a/.github/workflows/ci.yml
+++ b/.github/workflows/ci.yml
@@ -98,6 +98,33 @@ jobs:
           restore-keys: |
             spm-sourcepackages-${{ runner.os }}-xcode${{ env.XCODE_VERSION }}-
 
+      # ── Purge SDK-bound caches (#745) ────────────────────────────────────────
+      # The cached DerivedData restored above contains ModuleCache.noindex with
+      # precompiled `.pcm` module files (notably `SwiftShims-*.pcm`) that record
+      # the mtime of the SDK's `usr/lib/swift/shims/module.modulemap` they were
+      # built against. macos-15 runners and Xcode_16.4 patches across runs can
+      # have differing SDK file mtimes (cold-image refreshes, SDK shims relinked
+      # by Xcode), so the cached pcm refuses to load on the next run with:
+      #   error: file '.../SwiftShims-*.pcm' has been modified since the module
+      #   file '...' was built: mtime changed (was X, now Y)
+      # That trips xcodebuild with `exit 65` *before* tests run, so no
+      # `TestResults.xcresult` is produced — making the failure look like a
+      # source/test-compile bug (#746, #747) when the real cause is the SDK
+      # mtime drift documented in #745. Discarding both caches forces the
+      # compiler to rebuild them against the current runner's SDK while still
+      # keeping the much larger `Build/` intermediates (.swiftmodule, object
+      # files) around for incremental reuse — so the speedup from caching is
+      # preserved without the false failures.
+      - name: Purge SDK-bound module caches (avoid SwiftShims mtime drift)
+        run: |
+          for SUBDIR in ModuleCache.noindex SDKStatCaches.noindex; do
+            CACHE_PATH="${{ env.DERIVED_DATA_PATH }}/${SUBDIR}"
+            if [[ -d "$CACHE_PATH" ]]; then
+              echo "Removing stale ${SUBDIR} (would otherwise trigger SwiftShims mtime mismatch — see #745)"
+              rm -rf "$CACHE_PATH"
+            fi
+          done
+
       # ── Homebrew package cache ───────────────────────────────────────────────
       - name: Cache Homebrew packages
         id: brew-cache

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -124,6 +124,17 @@ to make the TCA Phase-1 reducers testable in isolation.
 
 ### 🛠 Internal / CI / Style
 
+- **#745 — Purge SDK-bound `ModuleCache.noindex` / `SDKStatCaches.noindex`
+  before each CI build.** The cached `DerivedData` restored across runs
+  contained precompiled `SwiftShims-*.pcm` files that recorded an SDK
+  `module.modulemap` mtime which drifts between macos-15 runner refreshes,
+  causing `xcodebuild` to fail with `error: file '.../SwiftShims-*.pcm'
+  has been modified since the module file '...' was built: mtime changed`
+  and `exit 65` *before* tests ran (no `TestResults.xcresult` produced).
+  Wiping just those two SDK-bound subdirectories preserves the much larger
+  `Build/` intermediates for incremental reuse while removing the false
+  failure surface that was being mis-attributed to source-compile bugs in
+  #746 / #747.
 - **#663 — Integrate SwiftLint into CI** per Google Swift Style.
 - **#685 — Line-wrap pass** in Views to comply with the 100-char column
   limit (closes #650).

